File: leetcode/new/leetcode/3sum-closest.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var threeSumClosest = function(nums, target) {
    nums.sort(function(a, b){return a-b})
    let curr_sum = nums.slice(0, 3).reduce((partialSum, a) => partialSum + a, 0);
    let out = curr_sum

    for(let i = 0; i < (nums.length - 2); i++){
        let left = i + 1;
        let right = nums.length - 1;
        while(left < right) {
            curr_sum = nums[left] + nums[i] + nums[right]
            if(Math.abs(curr_sum - target) < Math.abs(out - target)){
                out = curr_sum;
            }
            if (curr_sum > target){
                right--;
            } else {
                left++;
            } 
            if (curr_sum === target) {
                return target;
            }
        }
    }
    return out;
};

File: leetcode/3sum-closest.py
class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        curr_sum = sum(nums[:3])
        nums.sort()
        out = curr_sum
        for i in range(nums.__len__() - 2):
            left = i + 1
            right = len(nums) - 1
            while left < right:
                curr_sum = nums[i] + nums[left] + nums[right]
                if abs(curr_sum - target) < abs(out - target):
                    out = curr_sum
                if curr_sum == target:
                    return target
                if curr_sum > target:
                    right -= 1
                else:
                    left += 1

        return out